### ISSUE TITLE
Bug when provisioning MiniKube on AWS using miniKF, version 20201112.0.0

### ISSUE THREAD

**Issue by peemblis** [kubeflow/kubeflow#5423]
**What steps did you take and what happened:**
I'm following this guide: https://www.arrikto.com/kubeflow/news/minikf-on-aws/ 

Step 16 fails. I.e., after you SSH onto the EC2 instance and type in `minikf`.

I can trace the failure to line 74 of `update-preseed-aws.py`. In particular, this part:

```
primary_address = socket.gethostbyname(socket.gethostname())
```

NOTE: I was able to get it to continue to run by hardcoding the IP address into `update-preseed-aws.py`. It continues to run, then `provision.py` will fails as well because it contains the same code in line 2462. If I hardcode it there, it runs and eventually fail again (with a related error). It gets stuck here:
![image](https://user-images.githubusercontent.com/7805233/99857600-2d763900-2b51-11eb-9d29-a2f24a643d76.png)

If I `ctrl + c`, I see this:

```
2020-11-20T22:21:31.782117+0000 provision.py pid=19299/tid=19299/pytid=140582985160448 utils:114 [ERROR] ingress_health_check() failed: HTTPSConnectionPool(host='ip-xx-xxx-xx-xxx.eng.aws.abcde', port=443): Max retries exceeded with url: / (Caused by NewConnectionError('<requests.packages.urllib3.connection.HTTPSConnection object at 0x7fdbfc5a4400>: Failed to establish a new connection: [Errno -2] Name or service not known',)). Retrying in 1.0 secs.
```

**What did you expect to happen:**
I expected it to go to Step 17 (of the link above) with the prompt: "Provisioning Minikube..."

**Anything else you would like to add:** 

Below is the initial stack trace.
```
2020-11-20T20:17:51.132499+0000 update-preseed-aws.py pid=2149/tid=2149/pytid=139976716932864 cli:214 [CRITICAL] An unexpected exception has occured
2020-11-20T20:17:51.132659+0000 update-preseed-aws.py pid=2149/tid=2149/pytid=139976716932864 cli:215 [ERROR] [Errno -2] Name or service not known
Traceback (most recent call last):
  File "build/bdist.linux-x86_64/egg/rok_tasks/frontend/cli.py", line 196, in run
  File "/home/ubuntu/update-preseed-aws.py", line 74, in main
    primary_address = socket.gethostbyname(socket.gethostname())
socket.gaierror: [Errno -2] Name or service not known
```

I can use python3 to reproduce it:

```
ubuntu@ip-xx-xxx-xx-xxx:~$ python3
Python 3.5.2 (default, Jul 17 2020, 14:04:10) 
[GCC 5.4.0 20160609] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import socket
>>> socket.gethostname()
'ip-xx-xxx-xx-xxx'
>>> socket.gethostbyname(socket.gethostname())
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
socket.gaierror: [Errno -2] Name or service not known
```

**Environment:**
- MiniKF: 20201112.0.0
- Kubeflow version: (version number can be found at the bottom left corner of the Kubeflow dashboard):
- kfctl version: (use `kfctl version`):
- Kubernetes platform: (e.g. `minikube`)
- Kubernetes version: (use `kubectl version`):
- OS (e.g. from `/etc/os-release`):

**Comment by peemblis** [comment]
I've got a work around. Add the lines below to `/etc/hosts`:

```
xx.xxx.xx.xxx   ip-xx-xxx-xx-xxx
xx.xxx.xx.xxx   ip-xx-xxx-xx-xxx.eng.aws.abcde
```

where 

- `xx.xxx.xx.xxx` = external IP address 
- `ip-xx-xxx-xx-xxx` = `hostname` 
- `ip-xx-xxx-xx-xxx.eng.aws.abcde` = `fqdn`

**Comment by PatrickXYS** [comment]
/cc @kubeflow/arrikto 

Can you help check?

**Comment by tbaums** [comment]
Hey @peemblis - Thanks so much for filing the issue. I'm taking a look.

Can you confirm that you used the 'seller recommended' settings for subnets and security groups?

**Comment by tbaums** [comment]
Hey @peemblis - 

While we are at it, a few more questions:

**Comment by peemblis** [comment]
Thank you for your replies. 

For subnet, there are various subnets available to us in N. Virginia (us-east-1). I chose one that contained `private` in the name (as opposed to a subnet containing `public`). I'm afraid I cannot provide much additional information beyond this.

I used the recommended security group that is created as part of the tutorial.  

```
ubuntu@ip-xx-xxx-xx-xxx:~$ hostname
ip-xx-xxx-xx-xxx
ubuntu@ip-xx-xxx-xx-xxx:~$ host $(hostname)
Host ip-xx-xxx-xx-xxx not found: 3(NXDOMAIN)
ubuntu@ip-xx-xxx-xx-xxx:~$ host xx.xxx.xx.xxx
yyy.yy.yyy.yy.in-addr.arpa domain name pointer ip-xx-xxx-xx-xxx.ec2.internal.
```

**Comment by PatrickXYS** [comment]
/assign @tbaums 

Assign to unsubscribe me from thread

**Comment by vkoukis** [comment]
@peemblis Thanks for providing us with all this info. We have been deploying MiniKF on AWS multiple times during our internal testing, so we'd like to understand what is different in your specific networking configuration compared to the configurations we have been testing with.

If we manage to understand this difference, we will have solved your current issue.
So, here is what MiniKF *expects* to happen when configuring the machine.
Please run the following steps on *your* environment, and we will compare the results.
Please make sure to *undo* your workaround entries in `/etc/hosts` so they don't interfere with our exploration.

1. Get the hostname of the machine:
   ```
   $ hostname
   ip-172-31-XXX-XXX
   $ cat /etc/hostname
   ip-172-31-XXX-XXX
   ```

2. Perform a DNS query for this hostname, and get back your "primary" IP address. It may be an internal address.
   ```
   $ host $(hostname)
   ip-172-31-XXX-XXX.ec2.internal has address 172.31.XXX-XXX
   ```
   *This* is the step that is failing for you, and we need to understand why. This step depends on proper configuration of a default domain name in `/etc/resolv.conf`. So, let's have a look at it.

3. Here are the contents of my `/etc/resolv.conf` file:
   ```
   # Dynamic resolv.conf(5) file for glibc resolver(3) generated by resolvconf(8)
   #     DO NOT EDIT THIS FILE BY HAND -- YOUR CHANGES WILL BE OVERWRITTEN
   nameserver 172.31.0.2
   search ec2.internal
   ```
   Please do not redact important parts of this file. If there is a sensitive name, e.g., of a company, replace with `company`. But let's see the exact IP addresses. Note `172.31.0.2` *is* the internal address at which the AWS VPC DNS service listens at, see https://docs.aws.amazon.com/vpc/latest/userguide/vpc-dns.html#vpc-dns-hostnames

4. Perform a reverse lookup on the primary address, to get back the FQDN:
    ```
   $ host 172.31.XXX.XXX
   XXX.XXX.31.172.in-addr.arpa domain name pointer ip-172-31-XXX-XXX.ec2.internal.
   ```

5. Use this FQDN to construct the final URL to connect to.

We expect step (2) to fail, and something interesting to come up at step (3). Could you show your VPC networking settings?

6. Go to "Your VPCs" --> click on the VPC you have deployed into --> confirm "DNS hostnames" is enabled, "DNS resolution" is enabled. Click on the specific "DHCP options set" you'reusing.

7. What are the values for "domain-name" and "domain-name-servers"? The defaults are `ec2.internal` and `AmazonProvidedDNS`. But based on the information you have provided so far, it seems you are using some sort of private DNS infrastructure, which doesn't allow step (2) to complete successfully.

Finally, please note that being able to perform the above steps to determine the FQDN is a requirement for many software packages in general, so it's a good thing to get to the bottom of this.

Looking forward to your reply!

**Comment by peemblis** [comment]
Thanks, @vkoukis. 


First, I will undo the changes to `/etc/hosts` mentioned above. To confirm changes are made, I rerun `minikf` at the command line to reproduce the issue in the opening post: 

```
2020-11-24T14:17:38.112655+0000 update-preseed-aws.py pid=27714/tid=27714/pytid=140503181731584 cli:214 [CRITICAL] An unexpected exception has occured
2020-11-24T14:17:38.112776+0000 update-preseed-aws.py pid=27714/tid=27714/pytid=140503181731584 cli:215 [ERROR] [Errno -2] Name or service not known
Traceback (most recent call last):
  File "build/bdist.linux-x86_64/egg/rok_tasks/frontend/cli.py", line 196, in run
  File "/home/ubuntu/update-preseed-aws.py", line 74, in main
    primary_address = socket.gethostbyname(socket.gethostname())
socket.gaierror: [Errno -2] Name or service not known
An unexpected exception has occured
Line: /home/ubuntu/update-preseed-aws.py:74
Type: gaierror
Reason: [Errno -2] Name or service not known

The logfile `/home/ubuntu/update-preseed-aws.log' may contain more information on this error, including a backtrace.

Failed to update the MiniKF preseed file.
Aborting...
``` 
--------------------------------
I will now follow your steps:

1.  Hostname
```
ubuntu@ip-10-xxx-xx-xxx:~$ hostname
ip-10-xxx-xx-xxx
ubuntu@ip-10-160-xx-xxx:~$ cat /etc/hostname
ip-10-xxx-xx-xxx
```

2. DNS query on hostname

```
ubuntu@ip-10-xxx-xx-xxx:~$ host $(hostname)
Host ip-10-xxx-xx-xxx not found: 3(NXDOMAIN)
```

3.  Contents of `resolv.conf`

```
ubuntu@ip-10-xxx-xx-xxx:~$ cat /etc/resolv.conf
# Dynamic resolv.conf(5) file for glibc resolver(3) generated by resolvconf(8)
#     DO NOT EDIT THIS FILE BY HAND -- YOUR CHANGES WILL BE OVERWRITTEN
nameserver 10.xxx.xx.x
search eng.aws.abcde
```

4. Reverse lookup FQDN

```
ubuntu@ip-10-xxx-xx-xxx:~$host 10.xxx.xx.xxx
165.xx.xxx.xx.in-addr.arpa domain name pointer ip-10-xxx-xx-xxx.ec2.internal.
```

5. Use FQDN to construct URL (did I need to do anything here?)

6. VPC Settings
- DNS hostnames is Enabled
- DNS resolution is Enabled

7. DHCP options setup
- domain-name: eng.aws.abcde
- domain-name-servers: AmazonProvidedDNS

**Comment by peemblis** [comment]
(With regard to 5.) Note that when provisioning has completed: 
![image](https://user-images.githubusercontent.com/7805233/100110305-686abc00-2e32-11eb-8e5b-54aac998a7fd.png)

However, I can't connect to the central dashboard using the URL provided in the screenshot. I can connect to it using the primary IP address: 10.xxx.xx.xxx.

**Comment by vkoukis** [comment]
@peemblis Thank you for the information. Based on what you have described, our current best understanding is that you are using custom [e.g., corporate] DNS configuration, which does not resolve the hostname of your EC2 instance properly. So, MiniKF cannot detect its fully-qualified domain name [FQDN, e.g., `ip-xxxx-xxx-xxx-xxx.ec2.internal`], so it cannot configure itself.

More details follow:

These are the contents of your `/etc/resolv.conf`:
```
nameserver 10.xxx.xx.x
search eng.aws.abcde
```
You have redacted the IP of the nameserver, so we can't be sure, but it should be ending in `.2`, and be the Amazon-provided DNS. Do you confirm? If not, can we see the IP?

The crux of the problem is that you are using a *custom* search domain, `eng.aws.abcde`, and not the default Amazon-provided one, e.g., `ec2.internal`. This domain does *not* include an entry for the new MiniKF instance, so we cannot resolve the hostname of the MiniKF instance properly:
```
ubuntu@ip-10-xxx-xx-xxx:~$ host $(hostname)
Host ip-10-xxx-xx-xxx not found: 3(NXDOMAIN)
```

In your next message, you talk about what happens when provisioning completes:
> However, I can't connect to the central dashboard using the URL provided in the screenshot. I can connect to it using the primary IP address: 10.xxx.xx.xxx.

I assume you use your `/etc/hosts`-based workaround to allow provisioning to complete.
This explains why the URL doesn't work: MiniKF can now compute its FQDN, because of the added entries in `/etc/hosts`, but this FQDN doesn't exist on the custom DNS server that you are using to serve your custom `eng.aws.abcde` zone. There is no `ip-10-xxx-xxx-xxx.eng.aws.abcde` entry. So, your own browser fails to resolve it, and that's why you can't connect to it. You can confirm by running `nslookup ip-xxx-xxx-xxx-xxxx.eng.aws.abcde` [Windows] or `host ip-xxx-xxx-xxx-xxx.eng.aws.abcde` [Linux, macOS].

Similarly, I assume you are using some sort of corporate VPN to connect to the internal `10.xxx.xxx.xxx` IP of the MiniKF instance, and this works.

To summarize, we cannot support this. Your instance *must* be able to compute its FQDN. In general, lots of software needs properly configured DNS so it can compute the FQDN of the node it runs on, this is a very common requirement.

Here are some potential solutions:
1. Deploy inside a VPC that is properly configured. Either use a public subnet, or use a private subnet without the problematic custom domain `eng.aws.abcde`.
1. Continue deploying inside this VPC, but fix your DNS infrastructure, so it resolves `ip-10-xxx-xxx-xxx.eng.aws.abcde`. You need to talk to the people who manage your DNS infrastructure for this.
1. Add entries to `/etc/hosts` inside the MiniKF VM, but note you won't be able to use the bogus FQDN that MiniKF computes in this case, you will have to connect to the private IP directly, as you are doing now.

I hope the above helps, looking forward to your feedback.

**Comment by peemblis** [comment]
Thanks for all your help, @vkoukis. 

I can confirm that the nameserver IP ends in `.2` and yes, I am on my company's VPN.

We rely on some private resources that are only available when using the private subnet (things like bitbucket/gitlab and artifactory) so a public subnet wouldn't work in my case.

----------------------------------------------------------------------

To provide a bit more information:

I have a **different ec2 instance** running in a private subnet on the same VPC. The python code (that was erroring on the MiniKF AMI) appears to resolve without error.

```
ubuntu@ip-10-xxx-xx-xxx:~$ python3
Python 3.6.9 (default, Oct  8 2020, 12:12:24) 
[GCC 8.4.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import socket
>>> socket.gethostname()
'ip-10-xxx-xx-xxx'
>>> socket.gethostbyname(socket.gethostname())
'10.xxx.xx.xxx'
```

In addition, on the **MiniKF AMI ec2 instance**, `socket.gethostname(socket.gethostname() + ".ec2.internal")` resolves without error.

```
ubuntu@ip-10-xxx-xx-xxx:~$ python3
Python 3.5.2 (default, Jul 17 2020, 14:04:10) 
[GCC 5.4.0 20160609] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import socket
>>> socket.gethostname()
'ip-10-xxx-xx-xxx'
>>> socket.gethostbyname(socket.gethostname())
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
socket.gaierror: [Errno -2] Name or service not known
>>> socket.gethostbyname(socket.gethostname() + '.ec2.internal')
'10.xxx.xx.xxx'
```

**Comment by vkoukis** [comment]
@peemblis Thanks for the extra info.

> We rely on some private resources that are only available when using the private subnet (things like bitbucket/gitlab and artifactory) so a public subnet wouldn't work in my case.

Please note we don't require that you work on a public subnet. We test MiniKF exhaustively with both public and private subnets, and actually we even test in air-gapped environments as well, where there is no access to the Internet.

However, we do require that you work with a DNS configuration for your VPC/subnet that allows the machine to have an FQDN.

>I have a different ec2 instance running in a private subnet on the same VPC. The python code (that was erroring on the MiniKF AMI) appears to resolve without error.

Then we have to understand what is different. Can you re-run the process of https://github.com/kubeflow/kubeflow/issues/5423#issuecomment-733014110 for the machine that works? Please also show the contents of `/etc/hosts`, and note I have included an extra (2a) step. To summarize:

0. Any workarounds / special entries in `/etc/hosts`?
1. Hostname
2. DNS query on hostname
2a. [an extra step] try `host ip-xxx-xxx-xxx-xxx.eng.aws.abcde` directly
3. Contents of `/etc/resolv.conf`
4. Reverse lookup FQDN

I think you will see that step (2) will succeed on your other instance, while it fails on MiniKF.
This happens either
* because of step (2a): Your custom DNS zone actually has an entry for `ip-xxx-xxx-xxx-xxx`, while it doesn't have one for MiniKF, *or*
* because of step (3): You have a different `/etc/resolv.conf`, which includes `ec2.internal` as a search domain.

Finally, on this:
> In addition, on the MiniKF AMI ec2 instance, socket.gethostname(socket.gethostname() + ".ec2.internal") resolves without error.

Yes, this is expected. Amazon manages the `.ec2.internal` zone and it makes sure that it has an `ip-xxx-xxx-xxx-xxx.ec2.internal` entry for `ip-xxx-xxx-xxx-xxx`. But your `domain-name` DHCP options in the VPC do not include `.ec2.internal` as a search domain.

You can talk to your VPC administators so they add `.ec2.internal` as a search domain, based on this:
https://docs.aws.amazon.com/vpc/latest/userguide/VPC_DHCP_Options.html#DHCPOptionSets
But you/they must also be aware of this warning in the docs:
> *Important:*
Some Linux operating systems accept multiple domain names separated by spaces. However, other Linux operating systems and Windows treat the value as a single domain, which results in unexpected behavior. If your DHCP options set is associated with a VPC that has instances with multiple operating systems, specify only one domain name.

To summarize: Could you get some feedback from your network administrators based on all the information we have presented in this issue? This seems like something you need to fix on your part.

**Comment by cloudbow** [comment]
I am getting this error as well. Trying the workaround with public subnet

**Comment by cloudbow** [comment]
What worked for me is the addition of  public ip to the /etc/hosts. 

But now I get a strange behaviour. Whenever somebody creates a notebook server it automatically adds the following to the yaml.

NaNGi

Anyone know why.

I have to remove this in aws to get things working.

**Comment by kimwnasptd** [comment]
I'm closing this issue, since the @vkoukis has provided some solutions and we need more feedback to further troubleshoot.

> But now I get a strange behaviour. Whenever somebody creates a notebook server it automatically adds the following to the yaml.
NaNGi

@cloudbow this is a distinct issue on its own. Could you open a new one so we can further discuss it there?

**Comment by omerfsen-gsnd** [comment]
I found the reason for this it is the dummy0 interface somehow overlaps. So I modified it to be on different network (say 20.10.10 instead of default one )

so edit:


```bash
cat /etc/network/interfaces.d/99-minikf.cfg 

# https://justus.berlin/2018/08/dummy-network-device-in-ubuntu-linux-18-04/
auto dummy0
iface dummy0 inet static
      address 10.10.10.10
      netmask 255.255.255.0
      pre-up modprobe dummy numdummies=1
      post-down rmmod dummy

```

to 

```bash
cat /etc/network/interfaces.d/99-minikf.cfg 

# https://justus.berlin/2018/08/dummy-network-device-in-ubuntu-linux-18-04/
auto dummy0
iface dummy0 inet static
      address 20.10.10.10
      netmask 255.255.255.0
      pre-up modprobe dummy numdummies=1
      post-down rmmod dummy

```

and reboot.

**Comment by omerfsen-gsnd** [comment]
Also i have found out that even if you set dns reverse (PTR) query returns you standart domain that is <regio>.compute.internal 


```bash

dig -x 10.10.63.179
179.63.10.10.in-addr.arpa. 600  IN      PTR     ip-10-10-63-179.eu-west-1.compute.internal.

```
